src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -123,38 +131,60 @@ private function registerHandlers(ContainerBuilder $container)
                         $definitions[$serviceId = '.messenger.method_on_object_wrapper.'.ContainerBuilder::hash($messageClass.':'.$messagePriority.':'.$serviceId.':'.$method)] = $wrapperDefinition;
                     }
 
-                    $handlersByMessage[$messageClass][$messagePriority][] = new Reference($serviceId);
+                    foreach ($handlerBuses as $handlerBus) {
+                        $handlersByBusAndMessage[$handlerBus][$messageClass][$messagePriority][] = $serviceId;
+                    }
                 }
             }
         }
 
-        foreach ($handlersByMessage as $message => $handlers) {
-            krsort($handlersByMessage[$message]);
-            $handlersByMessage[$message] = array_merge(...$handlersByMessage[$message]);
+        foreach ($handlersByBusAndMessage as $bus => $handlersByMessage) {
+            foreach ($handlersByMessage as $message => $handlersByPriority) {
+                krsort($handlersByPriority);
+                $handlersByBusAndMessage[$bus][$message] = array_unique(array_merge(...$handlersByPriority));
+            }
         }
 
-        $handlersLocatorMapping = array();
-        foreach ($handlersByMessage as $message => $handlers) {
-            if (1 === \count($handlers)) {
-                $handlersLocatorMapping['handler.'.$message] = current($handlers);
-            } else {
-                $d = new Definition(ChainHandler::class, array($handlers));
-                $d->setPrivate(true);
-                $serviceId = hash('sha1', $message);
-                $definitions[$serviceId] = $d;
-                $handlersLocatorMapping['handler.'.$message] = new Reference($serviceId);
+        $handlersLocatorMappingByBus = array();
+        foreach ($buses as $bus) {
+            $handlersLocatorMappingByBus[$bus] = array();
+        }
+        foreach ($handlersByBusAndMessage as $bus => $handlersByMessage) {
+            foreach ($handlersByMessage as $message => $handlersIds) {
+                if (1 === \count($handlersIds)) {
+                    $handlersLocatorMappingByBus[$bus]['handler.'.$message] = new Reference(current($handlersIds));
+                } else {
+                    $chainHandler = new Definition(ChainHandler::class, array(array_map(function (string $handlerId): Reference {
+                        return new Reference($handlerId);
+                    }, $handlersIds)));
+                    $chainHandler->setPrivate(true);
+                    $serviceId = hash('sha1', $bus.$message);
```

Comment:
I know it was done like that but... it needs to be changed. Can you prefix it the name of the service with `.messenger.chain_handler.` and use `ContainerBuilder::hash` instead of `hash` directly?